### PR TITLE
MCLOUD-5968: PostDeployCest is red

### PR DESCRIPTION
**Pull request by BaDos**
### Description

There are 2 PRs with red PostDeployCest:
https://github.com/magento/ece-tools/pull/728
https://github.com/magento/ece-tools/pull/724

This test checks that the post-deploy phase is not run if the deploy phase failed.
The small investigation shows that the flag FLAG_DEPLOY_HOOK_IS_FAILED is not created.

Added configured actions which will be run on fail.
Using these actions developers can create some flags, clean up some dirs, etc.

### Fixed Issues (if relevant)

https://jira.corp.magento.com/browse/MCLOUD-5968

### Manual testing scenarios

Covered with FT

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: src/OnFail/Action/SkipAction.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        $this->logger = $logger;
+        $this->actionName = $actionName;
+    }
+
+    /**
+     * Logs the information about action skipping
+     *
+     * @return void
+     */
+    public function execute()
```
Comment (BarnyShergold) [inline src/OnFail/Action/SkipAction.php]: Missing :void return type declaration

File: src/Scenario/Collector/Action.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+     *
+     * @param array $action
+     * @return array
+     */
+    public function collect(array $action): array
+    {
+        $actionData = [
+            'name' => $action['@name'],
+            'type' => $action['@type'] ?? '',
+            'skip' => isset($action['@skip']) && $action['@skip'] === 'true'
```
Comment (BarnyShergold) [inline src/Scenario/Collector/Action.php]: Missing comma on last element

**Review by BarnyShergold — CHANGES_REQUESTED** [review]
Please add description from https://jira.corp.magento.com/browse/MCLOUD-5968 to ticket description

File: src/OnFail/Action/ActionInterface.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * Action represent small action on fail such us create some flags, etc
+ */
+interface ActionInterface
+{
+    /**
+     * Executes the action.
+     *
+     * @return void
+     */
+    public function execute(); // The :void return type declaration that should be here would cause a BC issue
```
Comment (shiftedreality) [inline src/OnFail/Action/ActionInterface.php]: This is a new interface, please add `:void`

File: src/OnFail/Action/CreateDeployFailedFlag.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+     */
+    private $flagManager;
+
+    public function __construct(FlagManager $flagManager)
+    {
+        $this->flagManager = $flagManager;
+    }
+
+    /**
+     * @throws \Magento\MagentoCloud\Filesystem\Flag\ConfigurationMismatchException
```
Comment (shiftedreality) [inline src/OnFail/Action/CreateDeployFailedFlag.php]: Use short name

File: src/Scenario/Collector/Action.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    public function collect(array $action): array
+    {
+        $actionData = [
+            'name' => $action['@name'],
+            'type' => $action['@type'] ?? '',
+            'skip' => isset($action['@skip']) && $action['@skip'] === 'true',
+        ];
+
+        if (isset($action['@priority'])) {
+            $actionData['priority'] = (int) $action['@priority'];
```
Comment (shiftedreality) [inline src/Scenario/Collector/Action.php]: There probably should not be space after `(int)`

File: src/Scenario/Merger.php [hunk 6]
```diff
@@ -44,61 +52,93 @@ class Merger
...
-        $isSkipped = isset($step['@skip']) && $step['@skip'] === 'true';
-
-        $requiredAttributes = $isSkipped ? ['@name'] : self::$stepRequiredArgs;
+        $isSkipped = isset($item['@skip']) && $item['@skip'] === 'true';
+        $requiredAttributes = $isSkipped ? ['@name'] : self::$requiredArgs;
 
-        if ($missedArgs = array_diff($requiredAttributes, array_keys($step))) {
+        if ($missedArgs = array_diff($requiredAttributes, array_keys($item))) {
             throw new ValidationException(sprintf(
                 'Argument(s) "%s" are missed from step',
```
Comment (shiftedreality) [inline src/Scenario/Merger.php]: `...missed from step` probably not correct here

File: src/Scenario/Resolver.php [hunk 7]
```diff
@@ -43,16 +45,68 @@ public function __construct(ContainerInterface $container, Sorter $sorter)
...
+    }
 
-        foreach ($scenarios as $step) {
+    /**
+     * Creates Step instances
+     *
+     * @param array $steps
+     * @return array
+     * @throws ValidationException
+     * @throws \Magento\MagentoCloud\App\ContainerException
```
Comment (shiftedreality) [inline src/Scenario/Resolver.php]: Use short namespace

File: src/Scenario/Resolver.php [hunk 8]
```diff
@@ -43,16 +45,68 @@ public function __construct(ContainerInterface $container, Sorter $sorter)
...
+        ];
+    }
+
+    /**
+     * Creates Action instances
+     *
+     * @param array $actions
+     * @return array
+     * @throws ValidationException
+     * @throws \Magento\MagentoCloud\App\ContainerException
```
Comment (shiftedreality) [inline src/Scenario/Resolver.php]: Short namespace

File: src/Scenario/Processor.php [hunk 9]
```diff
@@ -68,18 +75,18 @@ public function execute(array $scenarios): void
...
-
-            throw new ProcessorException(
-                $exception->getMessage(),
-                $exception->getCode(),
-                $exception
-            );
         } catch (Throwable $exception) {
-            $message = 'Unhandled error: ' . $exception->getMessage();
-
+            $message = ($exception instanceof StepException ? '': 'Unhandled error: ') . $exception->getMessage();
```
Comment (shiftedreality) [inline src/Scenario/Processor.php]: What's the reason for this change?

File: src/Scenario/Merger.php [hunk 10]
```diff
@@ -44,61 +52,93 @@ class Merger
...
+                    $this->actionCollector->collect($action)
+                );
+            }
         }
 
         return $this->resolver->resolve($data);
     }
 
+    /**
+     * Normalizes items list.
```
Comment (shiftedreality) [inline src/Scenario/Merger.php:119]: Please add proper explanation

File: src/OnFail/Action/SkipAction.php [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace Magento\MagentoCloud\OnFail\Action;
+
+use Psr\Log\LoggerInterface;
+
+/**
+ * Class for skipped actions.
+ * Logs the information about skipped tests.
+ */
+class SkipAction implements ActionInterface
```
Comment (shiftedreality) [inline src/OnFail/Action/SkipAction.php:16]: What's the purpose of this?

**Review by shiftedreality — CHANGES_REQUESTED** [review]

File: src/Scenario/Processor.php (hunk 9 above)
Comment (BaDos) [reply src/Scenario/Processor.php]: Most of the changes were made to avoid code duplication.

File: src/OnFail/Action/SkipAction.php (hunk 11 above)
Comment (BaDos) [reply src/OnFail/Action/SkipAction.php:16]: If developers can add some custom actions, they should also be able to skip some actions.

File: src/OnFail/Action/ActionInterface.php [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+ * Action represent small action on fail such us create some flags, etc
+ */
+interface ActionInterface
+{
+    /**
+     * Executes the action.
+     *
+     * @return void
+     */
+    public function execute(): void;
```
Comment (NadiyaS) [inline src/OnFail/Action/ActionInterface.php:21]: There should be possibility to throw an exception, isn't it?

File: src/OnFail/Action/ActionInterface.php [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\MagentoCloud\OnFail\Action;
+
+/**
+ * Action represent small action on fail such us create some flags, etc
```
Comment (NadiyaS) [inline src/OnFail/Action/ActionInterface.php]: I would add "scenario action" to be more specific

File: src/Scenario/Processor.php (hunk 9 above)
Comment (shiftedreality) [reply src/Scenario/Processor.php]: The idea here is to catches non-expexted exceptions.

```
 } catch (StepException|ActionException $exception)
```

should help

File: src/OnFail/Action/SkipAction.php [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    /**
+     * Logs the information about action skipping
+     *
+     * @return void
+     */
+    public function execute(): void
+    {
+        $this->logger->info(sprintf('Step "%s" was skipped', $this->actionName));
```
Comment (NadiyaS) [inline src/OnFail/Action/SkipAction.php]: **Action** "%s" was skipped

File: src/OnFail/Action/SkipAction.php [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+ */
+declare(strict_types=1);
+
+namespace Magento\MagentoCloud\OnFail\Action;
+
+use Psr\Log\LoggerInterface;
+
+/**
+ * Class for skipped actions.
+ * Logs the information about skipped tests.
```
Comment (NadiyaS) [inline src/OnFail/Action/SkipAction.php]: > ... about skipped tests

maybe "skipped actions"?

File: src/OnFail/Action/CreateDeployFailedFlag.php [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+    /**
+     * Creates .deploy_is_failed flag.
+     *
+     * {@inheritDoc}
+     */
+    public function execute(): void
+    {
+        try {
+            $this->flagManager->set(FlagManager::FLAG_DEPLOY_HOOK_IS_FAILED);
```
Comment (NadiyaS) [inline src/OnFail/Action/CreateDeployFailedFlag.php:36]: This set() method also can throw ConfigurationMismatchException

**Review by shiftedreality — APPROVED** [review]

**Review by NadiyaS — APPROVED** [review]

**Comment by shiftedreality** [comment]
@BarnyShergold, please approve

**Review by BarnyShergold — APPROVED** [review]

**Comment by mveeramneni** [comment]
Testing covered by autotests, QA Approved
